guides/source/tuning_perf_for_deployment.md

```diff
@@ -0,0 +1,158 @@
+
+Rails' default settings are chosen for small applications. You can improve performance for your large application that serves a lot of requests by changing settings.
+
+This section contains common sense defaults based on the type and size of your application and the hosts on which it runs. You can improve performance more by testing your application specifically. See "Performance Testing" for details.
+
+These are production recommendations. Your development application will have different needs from a production application, and should use a different configuration.
+
+[Puma's deployment documentation](https://github.com/puma/puma/blob/master/docs/deployment.md) may also be useful.
+
+### Threads Per Process
+
+Rails uses 3 threads per process by default. A well-optimized I/O-heavy Rails application should specify 5 or 6 threads per process at maximum. Discourse, for example, benefits from about 5 threads per process. Discourse also executes many database queries per request and frequently uses Redis. More self-contained applications with fewer database and API queries benefit from around 3 threads per process.
+
+The default Puma configuration mentions that "as a rule of thumb, increasing the number of threads will increase how much traffic a given process can handle (throughput), but due to CRuby's Global VM Lock (GVL) it has diminishing returns and will degrade the response time (latency) of the application."
+
+To set the number of threads, you can change the call to the +threads+ method in +config/puma.rb+. Or you can set the +RAILS_MAX_THREADS+ environment variable, which will do the same.
+
+### Number of Processes
+
+When using hybrid threads and processes, you should run 1 process per available processor core. Automatic methods to determine the number of cores are unreliable. You should specify the number of processes manually.
+
+To set the number of worker processes, you can change the call to the +workers+ method in +config/puma.rb+. Or you can set the +WEB_CONCURRENCY+ environment variable, which will do the same.
+
+### Preloading
+
+Puma creates new workers from a master process. By loading your application code in the master process, you can avoid doing so after creating the worker. This permits sharing more memory across processes.
+
+In a few cases it may not make sense to preload your application. In that case it's possible to turn off application preloading.
+
+Puma preloads your application by default by calling +preload_app!+ in +config/puma.rb+. If you remove this call, your application will not be preloaded.
+
+### Memory Allocators and Configuration
+
+CRuby normally uses your system's default memory allocator. You can switch to another allocator such as [jemalloc](https://github.com/jemalloc/jemalloc) or [tcmalloc](https://github.com/google/tcmalloc). You can also configure your allocator &mdash; e.g. Linux's glibc malloc allows setting MALLOC_ARENA_MAX=1 to significantly reduce memory use.
```

Comment:
I always thought that a good default was MALLOC_ARENA_MAX=2, based on https://www.mikeperham.com/2018/04/25/taming-rails-memory-bloat/

Reply:
I was quoting Jean here. But looks like it can vary: https://devcenter.heroku.com/articles/tuning-glibc-memory-behavior

I'll change the wording to make it clear there's some room for variation.

Reply:
I only mentioned `MALLOC_ARENA_MAX` not its value. And yeah generally `2` is recommended.

In short it's because by default glibc use one arena per thread, so that's a lot of reserved memory for each thread they don't share. The idea is the lower it is, the more contention between threads, the higher it is, the more wasted memory. So it's an usual latency VS memory usage tradeoff.

Hence why `jemalloc` is generally prefered, as it tend to solve this issue without significantly degrading latency.